kubernetes/client/cache.go

```diff
@@ -0,0 +1,164 @@
+package client
+
+import "log"
+
+type objEntry struct {
+	metadata *V1ObjectMeta
+	obj      interface{}
+}
+
+// ObjectLister is a function that knows how to list objects.
+type ObjectLister func() ([]interface{}, string, error)
+
+// ObjectWatcher is a function that knows how to perform a watch.
+type ObjectWatcher func(resourceVersion string) (results <-chan *Result, errors <-chan error)
+
+// EventHandler is implemented by objects that want event notifications
+type EventHandler interface {
+	OnAdd(obj interface{})
+	OnUpdate(oldObj, newObj interface{})
+	OnDelete(obj interface{})
+}
+
+// Informer is an interface for things that can provide notifications
+type Informer interface {
+	AddEventHandler(handler EventHandler)
+}
+
+// Lister is an interface for things that can list objects for all namespaces or by namespace
+type Lister interface {
+	List() []interface{}
+	ByNamespace() []interface{}
+}
+
+// Cache is an implementation of a List/Watch cache
+type Cache struct {
+	Extractor        func(interface{}) *V1ObjectMeta
+	Lister           ObjectLister
+	Watcher          ObjectWatcher
+	allObjects       []objEntry
+	namespaceObjects map[string][]objEntry
+	eventHandlers    []EventHandler
+}
+
+func (c *Cache) AddEventHandler(handler EventHandler) {
+	c.eventHandlers = append(c.eventHandlers, handler)
+}
+
+func (c *Cache) Run() {
+	for {
+		if err := c.ListWatch(); err != nil {
```

Comment:
exponential backoff? otherwise there could be hot looping.

Reply:
added.